### PR TITLE
Add JDBC datasource

### PR DESCRIPTION
**Pull request by penghuo**
### Description
1. Add JDBC datasource support. Only support ``org.apache.hive.jdbc.HiveDriver``.
2. Doc. https://github.com/penghuo/os-sql/blob/pr/spark/jdbc/docs/user/ppl/admin/jdbc.rst#ppl-supported-for-jdbc-connector
 
### Issues Resolved
https://github.com/opensearch-project/sql/issues/1331
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/test/java/org/opensearch/sql/analysis/AnalyzerTestBase.java [hunk 1]
```diff
@@ -158,19 +176,24 @@ protected LogicalPlan analyze(UnresolvedPlan unresolvedPlan) {
 
   private class DefaultDataSourceService implements DataSourceService {
 
-    private StorageEngine storageEngine = storageEngine();
-    private final DataSource dataSource
-        = new DataSource("prometheus", DataSourceType.PROMETHEUS, storageEngine);
+    private final DataSource opensearchDataSource = new DataSource(DEFAULT_DATASOURCE_NAME,
+        DataSourceType.OPENSEARCH, storageEngine());
+    private final DataSource prometheusDataSource
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/analysis/AnalyzerTestBase.java:181]: Maybe: add a tests with another datasource too?

File: docs/user/ppl/admin/jdbc.rst [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+
+JDBC Connector Properties in DataSource Configuration
+=====================================================
+JDBC Connector Properties.
+
+* ``url`` [Required].
+    * This parameters provides the URL to connect to a database instance provided endpoint.
+* ``driver`` [Required].
+    * This parameters provides the Driver to connect to a database instance provided endpoint. The only supported ``org.apache.hive.jdbc.HiveDriver``
```
Comment (Yury-Fridlyand) [inline docs/user/ppl/admin/jdbc.rst:27]: Why only `HiveDriver` supported? There is no hardcoded restriction. Can I use anything else, for example, put `SQLite` driver into `CLASSPATH` and use it?

File: docs/user/ppl/admin/jdbc.rst [hunk 3]
```diff
@@ -0,0 +1,76 @@
+.. highlight:: sh
+
+====================
+Prometheus Connector
```
[the remark this answers is not in the document]
Comment (penghuo) [reply docs/user/ppl/admin/jdbc.rst:4]: Done.

File: docs/user/ppl/admin/jdbc.rst [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    }]
+
+PPL supported for jdbc connector
+================================
+
+JDBC Table Function
+-------------------
+JDBC datasource could execute direct SQL against the target database. The SQL must be supported by target database.
+
+Sample Example::
```
[the remark this answers is not in the document]
Comment (penghuo) [reply docs/user/ppl/admin/jdbc.rst:67]: Done.

File: doctest/build.gradle [hunk 5]
```diff
@@ -103,41 +157,9 @@ stopOpenSearch.finalizedBy stopPrometheus
...
-    opensearch_build += '-' + version_tokens[1]
-}
-String mlCommonsRemoteFile = 'https://ci.opensearch.org/ci/dbc/distribution-build-opensearch/' + opensearch_no_snapshot + '/latest/linux/x64/tar/builds/opensearch/plugins/opensearch-ml-' + opensearch_build + '.zip'
-String mlCommonsPlugin = 'opensearch-ml'
-
 testClusters {
     docTestCluster {
         keystore 'plugins.query.federation.datasources.config', new File("$projectDir/datasource", 'datasources.json')
-        // Disable loading of `ML-commons` plugin, because it might be unavailable (not released yet).
-        /*
```
[the remark this answers is not in the document]
Comment (penghuo) [reply doctest/build.gradle:126]: Try to remove uncessary code, anyway github have history.

File: jdbc/src/main/java/org/opensearch/sql/jdbc/functions/JDBCFunction.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  protected class JDBCFunctionTable implements Table {
+    /**
+     * return empty map at query analysis stage.
+     */
+    @Override
+    public Map<String, ExprType> getFieldTypes() {
+      return ImmutableMap.of();
+    }
+
+    // todo, the implement interface should be removed. create an issue....
```
[the remark this answers is not in the document]
Comment (penghuo) [reply jdbc/src/main/java/org/opensearch/sql/jdbc/functions/JDBCFunction.java:78]: tracking with issue. https://github.com/opensearch-project/sql/issues/1463

File: jdbc/src/main/java/org/opensearch/sql/jdbc/functions/JDBCFunction.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  @VisibleForTesting
+  protected class JDBCFunctionTableScanBuilder extends TableScanBuilder {
+
+    @Override
+    public TableScanOperator build() {
+      return new JDBCQueryOperator(sqlQuery, properties);
+    }
+
+    /**
+     * PPL by default add LogicalProject operator. It should be ignored.
```
[the remark this answers is not in the document]
Comment (penghuo) [reply jdbc/src/main/java/org/opensearch/sql/jdbc/functions/JDBCFunction.java:103]: Done.

File: jdbc/src/main/java/org/opensearch/sql/jdbc/functions/JDBCTableFunctionResolver.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      DataSourceMetadata dataSourceMetadata, PropertiesParser propertiesParser) {
+    this.dataSourceMetadata = dataSourceMetadata;
+    this.propertiesParser = propertiesParser;
+  }
+
+  @Override
+  public Pair<FunctionSignature, FunctionBuilder> resolve(FunctionSignature functionSignature) {
+    FunctionBuilder functionBuilder =
+        (properties, arguments) -> {
+          if (arguments.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (penghuo) [reply jdbc/src/main/java/org/opensearch/sql/jdbc/functions/JDBCTableFunctionResolver.java:47]: Done. Remove unnecessary check and catch ExpressionEvaluationException.

File: jdbc/src/main/java/org/opensearch/sql/jdbc/operator/JDBCQueryOperator.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  @Override
+  public boolean hasNext() {
+    return handleSQLException(() -> jdbcResponseHandle.hasNext());
+  }
+
+  @Override
+  public ExprValue next() {
+    return handleSQLException(() -> jdbcResponseHandle.next());
```
[the remark this answers is not in the document]
Comment (penghuo) [reply jdbc/src/main/java/org/opensearch/sql/jdbc/operator/JDBCQueryOperator.java:105]: Done.

File: jdbc/src/main/java/org/opensearch/sql/jdbc/functions/JDBCTableFunctionResolver.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+  @VisibleForTesting
+  public static final FunctionSignature JDBC_FUNCTION_SIGNATURE =
+      new FunctionSignature(JDBC_FUNCTION_NAME, List.of(STRING));
+
+  private final DataSourceMetadata dataSourceMetadata;
+
+  private final PropertiesParser propertiesParser;
+
+  public JDBCTableFunctionResolver(
```
[the remark this answers is not in the document]
Comment (penghuo) [reply jdbc/src/main/java/org/opensearch/sql/jdbc/functions/JDBCTableFunctionResolver.java:37]: Done.

File: jdbc/src/main/java/org/opensearch/sql/jdbc/operator/JDBCResultSetResponseHandle.java [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+      case Types.TIMESTAMP:
+        return Optional.of(TIMESTAMP);
+
+      // we assume the result is json encoded string. refer https://docs.cloudera.com/HDPDocuments/HDP2/HDP-2.0.0.2/ds_Hive/jdbc-hs2.html,
+      case Types.ARRAY:
+      case Types.JAVA_OBJECT:
+      case Types.STRUCT:
+        return Optional.of(STRING);
+      default:
+        return Optional.empty();
```
[the remark this answers is not in the document]
Comment (penghuo) [reply jdbc/src/main/java/org/opensearch/sql/jdbc/operator/JDBCResultSetResponseHandle.java:183]: Remove Optional.

File: docs/user/ppl/admin/jdbc.rst (hunk 2 above)
Comment (penghuo) [reply docs/user/ppl/admin/jdbc.rst]: Why only HiveDriver supported? There is no hardcoded restriction. 
> Yes, only hive supported. Explain the possible value in doc.

>  Can I use anything else, for example, put SQLite driver into CLASSPATH and use it?
You can, if you put SQLLite into classpath. But it is not prebuild.

File: core/src/test/java/org/opensearch/sql/analysis/AnalyzerTestBase.java (hunk 1 above)
Comment (penghuo) [reply core/src/test/java/org/opensearch/sql/analysis/AnalyzerTestBase.java]: Any test case should be covered in analyer? Analyzer should not differenate different datasource.

**Review by Yury-Fridlyand — APPROVED** [review]

**Review by MaxKsyunz — COMMENTED** [review]
~My main question is: can this data source be used with any JDBC driver?~
~If not, is it possible to make it clear to the users that Spark is the only supported JDBC data source?~
I see now this was addressed in docs section.

I also had a question about if the change to security policies is required?

As a minor point, I flagged a few instance where `Immutable*` types were used instead of Java core types.

**Comment by luyuncheng** [comment]
why we close this pr? is there any continues planning for this?

**Comment by nazarovkv** [comment]
> may I ask about why we close this pr? is there any continues planning for this?

+1 

@penghuo is it possible to proceed with changes? Looks like community can loose valuable feature.

**Pull request closed without merging**